aten/src/ATen/native/cuda/ConvolutionMM2d.cu

```diff
@@ -1,12 +1,525 @@
+#include <ATen/cuda/CUDABlas.h>
+#include <ATen/native/ConvUtils.h>
+#include <ATen/native/Resize.h>
+#include <ATen/native/cuda/im2col.cuh>
 
-std::tuple<Tensor&, Tensor&, Tensor&> slow_conv2d_backward_out_cuda(const Tensor& grad_output,
-    const Tensor& self,
-    const Tensor& weight,
+namespace at { namespace native {
+namespace {
+
+void slow_conv2d_shape_check(
+    const Tensor& input, const Tensor& grad_output,
+    const Tensor& weight, const Tensor& bias,
+    int64_t kH, int64_t kW,
+    int64_t dH, int64_t dW,
+    int64_t padH, int64_t padW,
+    bool weight_nullable) {
+  TORCH_CHECK(kW > 0 && kH > 0,
+              "kernel size should be greater than zero, but got kH: ", kH, " kW: ", kW);
+  TORCH_CHECK(dW > 0 && dH > 0,
+              "stride should be greater than zero, but got dH: ", dH, " dW: ", dW);
+
+  TORCH_CHECK(weight_nullable || weight.defined(),
+              "weight tensor is expected to be non-nullable");
+  TORCH_CHECK(!weight.defined() ||
+              ((weight.numel() > 0) && (weight.dim() == 2 || weight.dim() == 4)),
+              "non-empty 2D or 4D weight tensor expected, but got: ", weight.sizes());
+  TORCH_CHECK(!bias.defined() || (bias.dim() == 1 && bias.sizes()[0] == weight.sizes()[0]),
+              "Expected bias to have shape [", weight.sizes()[0], "] but got ", bias.sizes());
+
+  const auto in_sizes = input.sizes();
+  int ndim = in_sizes.size();
+  int dimf = 0;
+  int dimh = 1;
+  int dimw = 2;
+
+  if (ndim == 4) {
+    dimf++;
+    dimh++;
+    dimw++;
+  }
+
+  // Allow for empty batch size but not other dimensions
+  const bool valid_empty = c10::multiply_integers(in_sizes.slice(1)) != 0;
+  TORCH_CHECK((valid_empty) && (ndim == 3 || ndim == 4),
+              "non-empty 3D or 4D input tensor expected but got: ", in_sizes);
+
+  int64_t inputHeight = in_sizes[dimh];
+  int64_t inputWidth = in_sizes[dimw];
+
+  int64_t exactInputHeight = inputHeight + 2 * padH;
+  int64_t exactInputWidth = inputWidth + 2 * padW;
+
+  TORCH_CHECK(exactInputHeight >= kH && exactInputWidth >= kW,
+              "Calculated padded input size per channel: ",
+              IntArrayRef{exactInputHeight, exactInputWidth},
+              ". Kernel size: ", IntArrayRef{kH, kW},
+              ". Kernel size can't be greater than actual input size");
+
+  // NOTE: can't use conv_output_size if the weight isn't defined
+  auto outputHeight = div_rtn<int64_t>(exactInputHeight - kH, dH) + 1;
+  auto outputWidth = div_rtn<int64_t>(exactInputWidth - kW, dW) + 1;
+
+  TORCH_CHECK(outputWidth >= 1 && outputHeight >= 1,
+              "Given input size per channel: ",
+              IntArrayRef{inputHeight, inputWidth},
+              ". Calculated output size per channel: ",
+              IntArrayRef{outputHeight, outputWidth},
+              ". Output size is too small");
+
+  if (weight.defined()) {
+    const auto w_sizes = weight.sizes();
+    int64_t nInputPlane = w_sizes[1];
+    if (w_sizes.size() == 2) {
```

Comment:
can this happen? I though weight shape is checked to have 4 dims before dispatching to slow conv?